Added MSA implementations for mips platforms.

mipsopen-fwu opened the pull request:
Signed-off-by: Fei Wu <fwu@wavecomp.com>

1. Implemented toolchain configurations for the build of mips32 and mips64 platforms under platform/linux/ directory. The mips toolchain can be foud from the link: https://www.mips.com/develop/tools/codescape-mips-sdk/ .
2. Implemented MIPS SIMD (MSA) intrinsics under modules/core/include/opencv2/core/ directory.
3. Made some optimizations with MSA functions in dnn/imgproc/video/videoio modules.

```
force_builders=linux,docs,Custom
buildworker:Custom=linux-1
build_image:Custom=mips64el
```

mipsopen-fwu commented:
I have added MSA implementations for MIPS platforms. The changes include:
1. Added toolchain configuration for mips 32 and 64 platform under platform/linux/.
2. Added intrinsics interfaces for MSA under modules/core/include/opencv2/core/.
3. Added some optimizations for MSA in dnn/imgproc/video/videoio modules.
Performance and functionality testsuites have been run locally for verification.

alalek commented on modules/dnn/src/layers/convolution_layer.cpp:
```diff
@@ -1020,6 +1020,17 @@ class ConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
+                                    vs12 = v_fma(w1, r2, vs12);
+                                    vs13 = v_fma(w1, r3, vs13);
+#else
                                     vs00 += w0*r0;
```
Perhaps this can be replaced to `v_fma()` unconditionally for all platforms.

alalek commented on modules/imgproc/src/demosaicing.cpp:
```diff
@@ -538,6 +538,233 @@ class SIMDBayerInterpolator_8u
+        {
+            v8u16 r0 = msa_ld1q_u16((const ushort*)bayer);
+            v8u16 r1 = msa_ld1q_u16((const ushort*)(bayer + bayer_step));
+            v8u16 r2 = msa_ld1q_u16((const ushort*)(bayer + bayer_step * 2));
```
Main principle of "universal intrinsics" (SIMD) approach is "code once, run anywhere".

You need to write code once (on some platform), debug it with some backend (there is even C++ backend for SIMD for reference) and then it should run on other backends with reasonable speed.

So we are strictly against huge code blocks using native intrinsic if this code can be replaced by universal SIMD code without significant lost of speed.
Universal SIMD code is regularly tested via x86/ARM backends, native RAW MIPS has very low testing coverage.

See discussion here: https://github.com/opencv/opencv/pull/9763#issuecomment-335159324

Could you please move these changes into a separate PR?
Lets merge first addition of SIMD backend and related build scripts (so we can add cross-platforms builds for that into our infrastructure (but no tests)).

mipsopen-fwu commented:
Hi Alexander

Thanks a lot for your quick response and detailed comments.
I will adjust and optimize the source code as you have commented, then submit separated PRs.

Best Regards,
Fei Wu

发件人: Alexander Alekhin <notifications@github.com>
发送时间: 2019年8月30日 19:31
收件人: opencv/opencv <opencv@noreply.github.com>
抄送: Fei Wu <fwu@wavecomp.com>; Author <author@noreply.github.com>
主题: [EXTERNAL]Re: [opencv/opencv] Added MSA implementations for mips platforms. (#15422)


@alalek commented on this pull request.

Great job 👍

General notes:

  *   please split PR: SIMD backend + build scripts should be reviewed and merged first. Other parts please move into separate PRs. Please check related comment below about RAW MSA code.
  *   as an "optimization" activity please re-target this PR onto 3.4 branch<https://github.com/opencv/opencv/wiki/Branches>

________________________________

In 3rdparty/libpng/CMakeLists.txt<https://github.com/opencv/opencv/pull/15422#discussion_r319438528>:

> @@ -46,6 +46,29 @@ if(";${CPU_BASELINE_FINAL};" MATCHES "SSE2"



   add_definitions(-DPNG_INTEL_SSE)



 endif()







+# set definitions and sources for MIPS



+if(CMAKE_SYSTEM_PROCESSOR MATCHES "mipsel*" OR



+   CMAKE_SYSTEM_PROCESSOR MATCHES "mips32*" OR



+   CMAKE_SYSTEM_PROCESSOR MATCHES "mips64*")



+  set(PNG_MIPS_MSA_POSSIBLE_VALUES on off)



+  set(PNG_MIPS_MSA "on" CACHE STRING "Enable MIPS_MSA optimizations:



+     off: disable the optimizations")



+  set_property(CACHE PNG_MIPS_MSA PROPERTY STRINGS



+     ${PNG_MIPS_MSA_POSSIBLE_VALUES})



-...



+set(PNG_MIPS_MSA ON CACHE BOOL ...)







-if(${PNG_MIPS_MSA} STREQUAL "on")



+if(PNG_MIPS_MSA)

instead of strings manipulation.

Also it is better to add this pre-check:

if(";${CPU_BASELINE_FINAL};" MATCHES ";MSA;")



________________________________

In cmake/OpenCVCompilerOptimizations.cmake<https://github.com/opencv/opencv/pull/15422#discussion_r319440242>:

> @@ -97,6 +98,8 @@ ocv_optimization_process_obsolete_option(ENABLE_FMA3 FMA3 ON)



 ocv_optimization_process_obsolete_option(ENABLE_VFPV3 VFPV3 OFF)



 ocv_optimization_process_obsolete_option(ENABLE_NEON NEON OFF)







+ocv_optimization_process_obsolete_option(ENABLE_MSA MSA OFF)



Please don't introduce obsolete options.
CPU_BASELINE=MSA should be used instead (it is defaulted to this values below on the line 350)

________________________________

In modules/core/include/opencv2/core/cvdef.h<https://github.com/opencv/opencv/pull/15422#discussion_r319441072>:

> @@ -256,6 +256,8 @@ namespace cv { namespace debug_build_guard { } using namespace debug_build_guard



 #define CV_CPU_AVX512_CEL       261



 #define CV_CPU_AVX512_ICL       262







+#define CV_CPU_MSA              300



Please use value "150"

________________________________

In modules/core/include/opencv2/core/msa_macros.h<https://github.com/opencv/opencv/pull/15422#discussion_r319441914>:

> @@ -0,0 +1,1558 @@



+// This file is part of OpenCV project.



+// It is subject to the license terms in the LICENSE file found in the top-level directory



+// of this distribution and at http://opencv.org/license.html.



+



+#ifndef OPENCV_CORE_HAL_MSA_MACROS_H



+#define OPENCV_CORE_HAL_MSA_MACROS_H



Please move this file into core/hal directory.

________________________________

In platforms/linux/mips.toolchain.cmake<https://github.com/opencv/opencv/pull/15422#discussion_r319443530>:

> +    set(CMAKE_MODULE_LINKER_FLAGS "" CACHE INTERNAL "")



+    set(CMAKE_EXE_LINKER_FLAGS    "-lpthread -lrt -ldl" CACHE INTERNAL "Added for mips cross build error")



+



+    set(CMAKE_CXX_FLAGS           "${CMAKE_CXX_FLAGS} -fdata-sections -Wa,--noexecstack -fsigned-char -Wno-psabi")



+    set(CMAKE_C_FLAGS             "${CMAKE_C_FLAGS} -fdata-sections -Wa,--noexecstack -fsigned-char -Wno-psabi")



+  endif()



+  set(CMAKE_SHARED_LINKER_FLAGS "${MIPS_LINKER_FLAGS} ${CMAKE_SHARED_LINKER_FLAGS}")



+  set(CMAKE_MODULE_LINKER_FLAGS "${MIPS_LINKER_FLAGS} ${CMAKE_MODULE_LINKER_FLAGS}")



+  set(CMAKE_EXE_LINKER_FLAGS    "${MIPS_LINKER_FLAGS} ${CMAKE_EXE_LINKER_FLAGS}")



+else()



+  #message(WARNING "CMAKE_CXX_FLAGS='${CMAKE_CXX_FLAGS}' is defined")



+endif()



+



+if(USE_MSA)



+  message(WARNING "You use obsolete variable USE_MSA to enable MSA instruction set. Use -DENABLE_MSA=ON instead." )



+  set(ENABLE_MSA TRUE)



Please remove obsolete options.

________________________________

In platforms/linux/mips.toolchain.cmake<https://github.com/opencv/opencv/pull/15422#discussion_r319443585>:

> +  set(CMAKE_SHARED_LINKER_FLAGS "${MIPS_LINKER_FLAGS} ${CMAKE_SHARED_LINKER_FLAGS}")



+  set(CMAKE_MODULE_LINKER_FLAGS "${MIPS_LINKER_FLAGS} ${CMAKE_MODULE_LINKER_FLAGS}")



+  set(CMAKE_EXE_LINKER_FLAGS    "${MIPS_LINKER_FLAGS} ${CMAKE_EXE_LINKER_FLAGS}")



+else()



+  #message(WARNING "CMAKE_CXX_FLAGS='${CMAKE_CXX_FLAGS}' is defined")



+endif()



+



+if(USE_MSA)



+  message(WARNING "You use obsolete variable USE_MSA to enable MSA instruction set. Use -DENABLE_MSA=ON instead." )



+  set(ENABLE_MSA TRUE)



+endif()



+



+set(CMAKE_FIND_ROOT_PATH ${CMAKE_FIND_ROOT_PATH} ${MIPS_LINUX_SYSROOT})



+



+if(EXISTS ${CUDA_TOOLKIT_ROOT_DIR})



+  set(CMAKE_FIND_ROOT_PATH ${CMAKE_FIND_ROOT_PATH} ${CUDA_TOOLKIT_ROOT_DIR})



Is there CUDA support on MIPS platform?

________________________________

In modules/core/src/matmul.simd.hpp<https://github.com/opencv/opencv/pull/15422#discussion_r319446645>:

> @@ -2385,6 +2385,48 @@ double dotProd_8u(const uchar* src1, const uchar* src2, int len)



             vst1q_u32(buf, v_sum);



             r += buf[0] + buf[1] + buf[2] + buf[3];







+            src1 += blockSize;



+            src2 += blockSize;



+            i += blockSize;



+        }



+    }



+#elif CV_MSA



+    if( cv::checkHardwareSupport(CV_CPU_MSA) )



check should be if (CV_CPU_HAS_SUPPORT_MSA) to call external dispatched SIMD code.
Anyway it doesn't have sense to add this runtime check this into dispatched code itself.

Please remove if() line. Compile-time check #if CV_MSA is enough.

BTW, similar NEON checks above are redundant too.

________________________________

In modules/core/include/opencv2/core/hal/intrin_msa.hpp<https://github.com/opencv/opencv/pull/15422#discussion_r319447510>:

> +}



+



+inline v_uint16x8 v_interleave_quads(const v_uint16x8& vec) { return v_reinterpret_as_u16(v_interleave_quads(v_reinterpret_as_s16(vec))); }



+



+inline v_int32x4 v_interleave_pairs(const v_int32x4& vec)



+{



+    v_int32x4 c;



+    c.val[0] = vec.val[0];



+    c.val[1] = vec.val[2];



+    c.val[2] = vec.val[1];



+    c.val[3] = vec.val[3];



+    return c;



+}



+



+inline v_uint32x4 v_interleave_pairs(const v_uint32x4& vec) { return v_reinterpret_as_u32(v_interleave_pairs(v_reinterpret_as_s32(vec))); }



+inline v_float32x4 v_interleave_pairs(const v_float32x4& vec) { return v_reinterpret_as_f32(v_interleave_pa



Please remove this call. It is gone (handled by compile time checks).

________________________________

In modules/core/include/opencv2/core/hal/intrin_msa.hpp<https://github.com/opencv/opencv/pull/15422#discussion_r319447859>:

> +}



+



+inline v_uint16x8 v_interleave_quads(const v_uint16x8& vec) { return v_reinterpret_as_u16(v_interleave_quads(v_reinterpret_as_s16(vec))); }



+



+inline v_int32x4 v_interleave_pairs(const v_int32x4& vec)



+{



+    v_int32x4 c;



+    c.val[0] = vec.val[0];



+    c.val[1] = vec.val[2];



+    c.val[2] = vec.val[1];



+    c.val[3] = vec.val[3];



+    return c;



+}



+



+inline v_uint32x4 v_interleave_pairs(const v_uint32x4& vec) { return v_reinterpret_as_u32(v_interleave_pairs(v_reinterpret_as_s32(vec))); }



+inline v_float32x4 v_interleave_pairs(const v_float32x4& vec) { return v_reinterpret_as_f32(v_interleave_pa



typo

________________________________

In 3rdparty/libpng/mips/filter_msa_intrinsics.c<https://github.com/opencv/opencv/pull/15422#discussion_r319456477>:

> +    out0 = (RTYPE) __msa_ilvr_b((v16i8) in0, (v16i8) in1);  \



+    out1 = (RTYPE) __msa_ilvr_b((v16i8) in2, (v16i8) in3);  \



+}



+#define ILVR_B2_SH(...) ILVR_B2(v8i16, __VA_ARGS__)



+



+#define HSUB_UB2(RTYPE, in0, in1, out0, out1)                 \



+{                                                             \



+    out0 = (RTYPE) __msa_hsub_u_h((v16u8) in0, (v16u8) in0);  \



+    out1 = (RTYPE) __msa_hsub_u_h((v16u8) in1, (v16u8) in1);  \



+}



+#define HSUB_UB2_SH(...) HSUB_UB2(v8i16, __VA_ARGS__)



+



+#define SLDI_B2_0(RTYPE, in0, in1, out0, out1, slide_val)                          \



+{                                                                                  \



+    out0 = (RTYPE) __msa_sldi_b((v16i8) __msa_fill_b(0), (v16i8) in0, slide_val);  \



+    out1 = (RTYPE) __msa_sldi_b((v16i8) __msa_fill_b(0), (v16i8) in1, slide_val);  \



Please use source from libpng 1.6.37
Add separate patch file if required (to avoid fixes lost after 3rdparty upgrade in the future)

________________________________

In modules/dnn/src/layers/convolution_layer.cpp<https://github.com/opencv/opencv/pull/15422#discussion_r319463039>:

> @@ -1020,6 +1020,17 @@ class ConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl



                                     v_float32x4 r0 = v_load_aligned(rptr), r1 = v_load_aligned(rptr + vsz_a),



                                                 r2 = v_load_aligned(rptr + vsz_a*2), r3 = v_load_aligned(rptr + vsz_a*3);







+#ifdef CV_MSA



+                                    vs00 = v_fma(w0, r0, vs00);



+                                    vs01 = v_fma(w0, r1, vs01);



+                                    vs02 = v_fma(w0, r2, vs02);



+                                    vs03 = v_fma(w0, r3, vs03);



+



+                                    vs10 = v_fma(w1, r0, vs10);



+                                    vs11 = v_fma(w1, r1, vs11);



+                                    vs12 = v_fma(w1, r2, vs12);



+                                    vs13 = v_fma(w1, r3, vs13);



+#else



                                     vs00 += w0*r0;



Perhaps this can be replaced to v_fma() unconditionally for all platforms.

________________________________

In modules/imgproc/src/demosaicing.cpp<https://github.com/opencv/opencv/pull/15422#discussion_r319469068>:

> +    int bayer2Gray(const uchar* bayer, int bayer_step, uchar* dst,



+                   int width, int bcoeff, int gcoeff, int rcoeff) const



+    {



+        /*



+         B G B G | B G B G | B G B G | B G B G



+         G R G R | G R G R | G R G R | G R G R



+         B G B G | B G B G | B G B G | B G B G



+         */



+        v8u16 masklo = msa_dupq_n_u16(255);



+        const uchar* bayer_end = bayer + width;



+



+        for( ; bayer <= bayer_end - 18; bayer += 14, dst += 14 )



+        {



+            v8u16 r0 = msa_ld1q_u16((const ushort*)bayer);



+            v8u16 r1 = msa_ld1q_u16((const ushort*)(bayer + bayer_step));



+            v8u16 r2 = msa_ld1q_u16((const ushort*)(bayer + bayer_step * 2));



Main principle of "universal intrinsics" (SIMD) approach is "code once, run anywhere".

You need to write code once (on some platform), debug it with some backend (there is even C++ backend for SIMD for reference) and then it should run on other backends with reasonable speed.

So we are strictly against huge code blocks using native intrinsic if this code can be replaced by universal SIMD code without significant lost of speed.
Universal SIMD code is regularly tested via x86/ARM backends, native RAW MIPS has very low testing coverage.

See discussion here: #9763 (comment)<https://github.com/opencv/opencv/pull/9763#issuecomment-335159324>

Could you please move these changes into a separate PR?
Lets merge first addition of SIMD backend and related build scripts (so we can add cross-platforms builds for that into our infrastructure (but no tests)).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/opencv/opencv/pull/15422?email_source=notifications&email_token=AM4VXTMIFYNXOCA6KU4YE23QHEAGVA5CNFSM4ISK27G2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDG7VKQ#pullrequestreview-281934506>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AM4VXTISCEYHAQ5HZWRE5GDQHEAGVANCNFSM4ISK27GQ>.

mipsopen-fwu commented:
@alalek Since I need to separate the PR into two, should I close this pull request first? 
 Next Steps I'm planning to do: 
1. Create development branch from 3.4 and commit the changes for SIMD backend+build scripts.
2. Submit pull request for the commits in Step 1.
3. After the PR in Step 2 is merged, commit the optimizations for other modules and create another PR.

alalek commented:
Lets keep basic part of MSA integration in this PR (at least to track conversation above).

You can "force push" commits in your branch to update PR ("mipsopen-fwu:msa-dev").

mipsopen-fwu commented:
@alalek
I'm trying to rebase this pull request from 3.4 by following the steps specified in:
https://github.com/opencv/opencv/wiki/Branches#rebase-pull-request-from-master-to-34
Conflicts are reported when I changed base branch to '3.4'.  
Need I resolve these conflicts first before I do the following steps?
rebase your commits from master onto 3.4 branch 
git checkout <your-branch>
(optional) create backup branch: git branch <your-branch>-backup 
git remote update upstream (assuming upstream is pointing to opencv/opencv GitHub repository)
git rebase -i --onto upstream/3.4 upstream/master
editor will be opened, check list of commits - there should be only your commits, save and exit
git push --force origin <your-branch> (assuming origin is pointing to <your-username>/opencv GitHub repository)

alalek commented:
> Need I resolve these conflicts 

Yes. You can resolve conflicts during rebase.

Before that you may enable diff3 conflict style for more information about changes: https://stackoverflow.com/questions/27417656/should-diff3-be-default-conflictstyle-on-git
Edit conflicts. Add file into index (`git add -i`). Then continue rebase for next commit (`git rebase --continue`).

To abort rebase process use `git rebase --abort`

alalek commented:
I rebased PR onto 3.4 branch - fetch changes from your fork.
Looks like GitHub conflicts are appear after switching the branch (they are not related to your patch - just between our 3.4/master branches)

Old saved commits are here: https://github.com/alalek/opencv/commits/pr15422_0903

mipsopen-fwu commented:
@alalek Thanks for rebasing! So what I need to do for the next step is to separate SIMD backend & build scripts from my previous commits, then commit and 'force push' to my branch (msa-dev) to update PR, right?

mipsopen-fwu commented:
@alalek I have separated MSA backend and build scripts from the old commit in this PR which has been rebased to 3.4 branch.

alalek commented:
I have added builder with cross-compilation from Ubuntu 18:04 x64:
- see [here](http://pullrequest.opencv.org/#/summary/opencv) "Custom" column
- looks like toolchain is not able to [detect properly cross-compiler](http://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/2589) (perhaps we can't use this one - we need to fix it or create new one)
- list of installed packages and corresponding scripts are [here](https://github.com/opencv-infrastructure/opencv-worker-config/tree/linux/images/mips64r6el)

mipsopen-fwu commented:
> I have added builder with cross-compilation from Ubuntu 18:04 x64:
> 
> see here "Custom" column
> looks like toolchain is not able to detect properly cross-compiler (perhaps we can't use this one - we need to fix it or create new one)
> list of installed packages and corresponding scripts are here
> 
> 
> Update:
> 
> adjusted some build options, but we still need to remove strange code block to unblock compiler detection
@alalek Can you build the project for mips platform successfully now? 
Looking from the cmake errors, it seems all the mips compiling options were not recognized. I think maybe you need to intall the mips linux gnu toolchain. 
Here is the link for the toolchain installer:
Online installer:  https://codescape.mips.com/sdk/v2.0.0k/mipssdk.v2.0.0k.linux-x64.x64.webinstall.run
Offline installer: https://codescape.mips.com/sdk/v2.0.0k/mipssdk.v2.0.0k.linux-x64.x64.offline.run
Usually we use cmake-gui to configure and generate the buid scripts by selecting /platform/linux/mips64r6el-gnu.toolchain.cmake .

I will remove the unused code block you refered and commit it.

mipsopen-fwu commented:
@alalek
I reviewed the build result. There are one failure and some warnings.

1. https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/22854/steps/Generate%20ABI%20dump/logs/log
The GCC parameters:
  gcc -fdump-translation-unit -fkeep-inline-functions -c -x c++ -fpermissive -w  -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winit-self -Wno-delete-non-virtual-dtor -Wno-comment -Wno-missing-field-initializers -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections -msse -msse2 -msse3 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG -DNDEBUG -DOPENCV_ABI_CHECK=1 "/tmp/iJM73d75Px/dump1.h"  -I/build/precommit_linux64/install/include -I/build/precommit_linux64/install/include/opencv -I/usr/include/vtk-6.0 -I/build/precommit_linux64/build

In file included from /tmp/iJM73d75Px/dump1.h:55:0:
/build/precommit_linux64/install/include/opencv2/core/hal/msa_macros.h:8:17: fatal error: msa.h: No such file or directory
 #include "msa.h"
                 ^
compilation terminated.
-- There is no -mmsa compiling option in the gcc parameters, instead -msse is used.

2.  Warnings: https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/2596/steps/compile%20release/logs/warnings%20%2815%29
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:315:11: warning: declaration of 'zero_m' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:315:11: warning: declaration of 'zero_m' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:315:11: warning: declaration of 'zero_m' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:338:11: warning: declaration of 'zero' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:338:11: warning: declaration of 'zero' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:338:11: warning: declaration of 'zero' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:338:11: warning: declaration of 'zero' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:315:11: warning: declaration of 'zero_m' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:315:11: warning: declaration of 'zero_m' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:315:11: warning: declaration of 'zero_m' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:338:11: warning: declaration of 'zero' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:338:11: warning: declaration of 'zero' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:338:11: warning: declaration of 'zero' shadows a previous local [-Wshadow]
/build/precommit_custom_linux/3.4/opencv/3rdparty/libpng/mips/filter_msa_intrinsics.c:338:11: warning: declaration of 'zero' shadows a previous local [-Wshadow]

--These warnings are introduced by the original filter_msa_intrinsics.c file from libpng-1.6.37. We didn't make any changes on it. If needed, I can make a fix.

cc1plus: warning: 'void* __builtin_memcpy(void*, const void*, long unsigned int)' forming offset [25, 32] is out of the bounds [0, 24] of object 'scharr' with type 'const int [6]' [-Warray-bounds]

--This warning is not introduced by my commit.

alalek commented on modules/core/include/opencv2/core/hal/intrin_msa.hpp:
```diff
@@ -0,0 +1,1760 @@
+}
+
+inline v_uint32x4 v_interleave_pairs(const v_uint32x4& vec) { return v_reinterpret_as_u32(v_interleave_pairs(v_reinterpret_as_s32(vec))); }
+inline v_float32x4 v_interleave_pairs(const v_float32x4& vec) { return v_reinterpret_as_f32(v_interleave_pa
```
GitHub code reference is buggy again - correct line is 1748 about hasSIMD128().

alalek replied:
Please remove `static inline bool hasSIMD128()` - it is not used anymore in OpenCV.

alalek commented:
Thank you for recommended build instructions!
I believe it make sense to add them into PR description and/or add comment in r6 toolchain file with SDK link + this PR link.

Currently on CI we prefer using cross-compilation mode through builtin Ubuntu compiler kits. We believe it is enough for basic validation builds (no tests). It is easy enough to integrate and upgrade dependencies from our side.

1. It is Linux x86-64 build. No `-mmsa` option is expected here.
Each header is included directly into ABI checker. Need to add guard in hal/msa_macros.h:
```
#ifdef __mips_msa
...
#endif
```

2. In 3rdparty/libpng/CMakeLists.txt near `add_definitions(-DPNG_MIPS_MSA_OPT=2)` add this line:
```
ocv_warnings_disable(CMAKE_C_FLAGS -Wshadow)
```

3. OK, I will try to check this with GCC-8 on other archs and find out how to workaround that.

mipsopen-fwu commented:
@alalek  I checked the failed build log for 'Android pack' : https://pullrequest.opencv.org/buildbot/builders/precommit_pack_android/builds/11090/steps/build%20sdk/logs/stdio

This error Error ‘ opcode not supported on this processor: mips32 (mips32) `pause' ’ is not introduced by my commit.

I think maybe  this error can be fixed by adding __mips_isa_rev check in modules/core/src/parallel_impl.cpp:

# elif defined __GNUC__ && defined __mips__
#   define CV_PAUSE(v) do { for (int __delay = (v); __delay > 0; --__delay) { asm volatile("pause" ::: "memory"); } } while (0)

------>

# elif defined __GNUC__ && defined __mips__  &&  __mips_isa_rev >= 3
#   define CV_PAUSE(v) do { for (int __delay = (v); __delay > 0; --__delay) { asm volatile("pause" ::: "memory"); } } while (0)

mipsopen-fwu commented:
@alalek I have committed the fix for compiling error in 'Android pack' build

mipsopen-fwu commented:
Hi, Alexander

The PR #15422 is kept in ‘waiting for builds’ state currently. How can I make it go ahead?
I’ll submit another PR for the other parts of MSA optimizations after 15422 is merged.

Thanks.

Best Regards,
Fei Wu

发件人: Alexander Alekhin <notifications@github.com>
发送时间: 2019年9月6日 12:59
收件人: opencv/opencv <opencv@noreply.github.com>
抄送: Fei Wu <fwu@wavecomp.com>; Author <author@noreply.github.com>
主题: [EXTERNAL]Re: [opencv/opencv] Added MSA implementations for mips platforms. (#15422)


Thank you for recommended build instructions!
I believe it make sense to add them into PR description and/or add comment in r6 toolchain file with SDK link + this PR link.

Currently on CI we prefer using cross-compilation mode through builtin Ubuntu compiler kits. We believe it is enough for basic validation builds (no tests). It is easy enough to integrate and upgrade dependencies from our side.

  1.  It is Linux x86-64 build. No -mmsa option is expected here.
Each header is included directly into ABI checker. Need to add guard in hal/msa_macros.h:

#ifdef __mips_msa

...

#endif

  1.  In 3rdparty/libpng/CMakeLists.txt near add_definitions(-DPNG_MIPS_MSA_OPT=2) add this line:

ocv_warnings_disable(CMAKE_C_FLAGS -Wshadow)

  1.  OK, I will try to check this with GCC-8 on other archs and find out how to workaround that.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/opencv/opencv/pull/15422?email_source=notifications&email_token=AM4VXTJFXXMCNA6HTJRRTHDQIHPPRA5CNFSM4ISK27G2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6BXEDA#issuecomment-528708108>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AM4VXTJEXMOKMS37GT66CGTQIHPPRANCNFSM4ISK27GQ>.

alalek commented on cmake/OpenCVCompilerOptimizations.cmake:
```diff
@@ -339,6 +340,12 @@ elseif(ARM OR AARCH64)
+  ocv_update(CPU_MSA_TEST_FILE "${OpenCV_SOURCE_DIR}/cmake/checks/cpu_msa.cpp")
+  ocv_update(CPU_KNOWN_OPTIMIZATIONS "MSA")
+  ocv_update(CPU_MSA_FLAGS_ON "")
+  ocv_update(CPU_FP16_IMPLIES "MSA")
```
FP16 is not enabled in current builds.
What did you mean to make here?

alalek commented on modules/core/include/opencv2/core/cv_cpu_dispatch.h:
```diff
@@ -152,6 +152,11 @@
 
+#ifdef __mips_msa
+#  include "hal/msa_macros.h"
+#  define CV_MSA 1
```
It is better to perform check via `CV_CPU_COMPILE_MSA` (at least, for debug purposes of SIMD optimizations this would allow to disable MSA via CMake)

alalek commented on modules/core/include/opencv2/core/cvdef.h:
```diff
@@ -319,6 +321,8 @@ enum CpuFeatures {
     CPU_AVX512_CEL      = 261, //!< Cascade Lake with AVX-512F/CD/BW/DQ/VL/IFMA/VBMI/VNNI
     CPU_AVX512_ICL      = 262, //!< Ice Lake with AVX-512F/CD/BW/DQ/VL/IFMA/VBMI/VNNI/VBMI2/BITALG/VPOPCNTDQ
 
+    CPU_MSA             = 300,
```
150

mipsopen-fwu commented:
@alalek
1. Thanks a lot for committing patches for the '3rdparth' code.
2. FP16 related option seems to be some legacy code. Since it's not supported and mips also doesn't support 16 bit wide floating point register, I removed this line.
3. I replaced __mips_msa with CV_CPU_COMPILE_MSA for MSA feature checking in cv_cpu_dispatch.h.
4. CPU_MSA has been defined as 150 now.
5. hasSIMD128() has been removed.

mipsopen-fwu commented:
@terfendail
Sorry for late response.
1.  Yes, you are right.  CV_SIMD128_64F guarding is not needed since it's unconditionally supported in this implementation. I will removed them form intrin_msa.hpp, only the define is kept.
2.  The '#if 0' version is slower according to the test results. The reason is that MSA doesn't actually provide data type expanding instructions for vectors.  The 'movl' related conversion are done by C.
So it would take more steps to use 'movl' interface than doing converion directly.
3.  In case data width is less than wide_step * 2, the performance would be better with this change. Otherwise, the performance wouldn't be better. Considering 'width' should be larger in most cases, this change is not necessary. I will remove it.
4. I will remove CV_MSA related code block dotProd_8u(), it's not necessary.
Thanks.

mipsopen-fwu commented on modules/core/src/matmul.simd.hpp:
```diff
@@ -2390,6 +2390,45 @@ double dotProd_8u(const uchar* src1, const uchar* src2, int len)
             i += blockSize;
         }
     }
+#elif CV_MSA
```
I will remove CV_MSA related code block dotProd_8u(), it's not necessary.

mipsopen-fwu commented on modules/core/src/arithm.simd.hpp:
```diff
@@ -393,7 +393,7 @@ static void bin_loop(const T1* src1, size_t step1, const T1* src2, size_t step2,
     typedef bin_loader<OP, T1, Tvec> ldr;
     enum {wide_step = Tvec::nlanes};
-    #if !CV_NEON && CV_SIMD_WIDTH == 16
+    #if !CV_NEON && !CV_MSA && CV_SIMD_WIDTH == 16
```
In case data width is less than wide_step * 2, the performance would be better with this change. Otherwise, the performance shouldn't be better. Considering 'width' should be larger in most cases, this change would not be necessary. I will remove it.

mipsopen-fwu commented on modules/core/include/opencv2/core/hal/intrin_msa.hpp:
```diff
@@ -0,0 +1,1751 @@
+OPENCV_HAL_IMPL_MSA_EXPAND(v_uint32x4, v_uint64x2, uint, u32, s32, v4u32, v4i32)
+OPENCV_HAL_IMPL_MSA_EXPAND(v_int32x4, v_int64x2, int, s32, s32, v4i32, v4i32)
+
+#if 0
```
The '#if 0' version is slower according to the test results. The reason is that MSA doesn't actually provide data type expanding instructions for vectors. The 'movl' related conversion are done by C.
So it would take more steps to use 'movl' interface than doing converion directly.

mipsopen-fwu commented on modules/core/include/opencv2/core/hal/intrin_msa.hpp:
```diff
@@ -0,0 +1,1751 @@
+OPENCV_HAL_IMPL_MSA_CHECK_ALLANY(uint16x8, v8u16, u16, 15)
+OPENCV_HAL_IMPL_MSA_CHECK_ALLANY(uint32x4, v4u32, u32, 31)
+#if CV_SIMD128_64F
+OPENCV_HAL_IMPL_MSA_CHECK_ALLANY(uint64x2, v2u64, u64, 63)
```
Yes, you are right. CV_SIMD128_64F guarding is not needed since it's unconditionally supported in this implementation. I will removed them form intrin_msa.hpp, only the define is kept.

alalek commented on modules/core/src/matmul.simd.hpp:
```diff
@@ -2476,6 +2476,45 @@ double dotProd_8s(const schar* src1, const schar* src2, int len)
             i += blockSize;
         }
     }
+#elif CV_MSA
```
`#if CV_SIMD` blocks this code execution here too (and compilation). The same note is about `#elif CV_NEON` code block (dead code too)

terfendail replied:
@mipsopen-fwu Have you compared performance of this MSA specific implementation against generic UI implementation?

mipsopen-fwu replied:
@terfendail 
This CV_MSA related code block was added by mistake when I merged msa implementations from 4.0.1 branch to 4.1.1 and 3.4 branch.  On our pervious 4.0.1 development branch, this MSA specific implementation doesn't exist. It's correct to remove from the pull request.

Even though, in order to make clear the performance difference, I did MatType_Length_dot test  by setting testparam=(8UC1, 1024) and samples=15.  From the result, we can see that the performance of generic UI implementation(using CV_SIMD) is better against CV_MSA specific implementation which was added by mistake in this code block.
The following is the test result:
1) The  total running time for CV_MSA implementation is:
[----------] 18 tests from MatType_Length_dot (2177 ms total)
2) The total running time for CV_SIMD implementations is:
[----------] 18 tests from MatType_Length_dot (618 ms total)

mshabunin commented on cmake/OpenCVCompilerOptimizations.cmake:
```diff
@@ -339,6 +340,11 @@ elseif(ARM OR AARCH64)
+elseif(MIPS)
+  ocv_update(CPU_MSA_TEST_FILE "${OpenCV_SOURCE_DIR}/cmake/checks/cpu_msa.cpp")
+  ocv_update(CPU_KNOWN_OPTIMIZATIONS "MSA")
+  ocv_update(CPU_MSA_FLAGS_ON "")
```
Why is it empty? It should contain "-mmsa" or something similar.

mipsopen-fwu replied:
This option seems not be used elsewhere, but it should be better to be defined as '-mmsa' for future use. I will modify it.

mshabunin replied:
This option is used, but it not obvious:

* there is a list of `CPU_KNOWN_OPTIMIZATIONS` which contain `MSA`
* cmake first tries to compile file `CPU_${OPT}_TEST_FILE` with option `CPU_${OPT}_FLAGS_ON`, where `OPT=MSA`
* if compilation succeeded and `MSA` is requested by user (via `CPU_BASELINE`), then these options will be added to compiler

Is the `-mmsa` option the only one required by implemented intrinsics? What about `-fhard-float` or others? We should move these options from toolchain file to CompilerOptimization.cmake.

mipsopen-fwu replied:
Yes. -mmsa  option is enough. When '-mmsa' is included, '-mfp64' and '-mhard-float' will be enabled by default.  '-mmsa' has the same effect as '-mmsa -mfp64 -mhard-float'.

terfendail commented on modules/core/include/opencv2/core/hal/intrin_msa.hpp:
```diff
@@ -0,0 +1,1694 @@
+inline bool v_check_any(const v_float32x4& a)
+{ return v_check_any(v_reinterpret_as_u32(a)); }
+
+#if CV_SIMD128_64F
```
`v_check_all(const v_int64x2&)` shouldn't be guarded by CV_SIMD128_64F as well.

mipsopen-fwu replied:
I will remove CV_SIMD128_64F guarding form intrin_msa.hpp, only the define is kept.
